config/initializers/secure_headers.rb

```diff
@@ -33,3 +33,12 @@
   #   ]
   # }
 end
+
+SecureHeaders::Configuration.override(:saml) do |config|
+  providers = YAML.load_file("#{Rails.root}/config/service_providers.yml")
+  providers = providers.fetch(Rails.env, {})
+  providers.symbolize_keys!
+  whitelisted_urls = providers[:valid_hosts].values.map { |k, _| k['acs_url'] }
+
+  whitelisted_urls.each { |url| config.csp[:form_action] << url }
+end
```

Comment:
this could be improved, also. the only whitelisted URL needed for each request is the one associated with the requesting SP. rather than including all domains for the current env, just fetch the domain from the root of the ACS URL for the current requesting SP